arg_spec - move validator lookup method to a function

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Move the main method that looks up the validator out of `AnsibleModule` and into a standalone function.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/module_utils/basic.py`
`lib/ansible/module_utils/common/parameters.py`

samdoran committed:
arg_spec - move type checking lookup method to a function

Shrews commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1849,20 +1851,10 @@ def _handle_options(self, argument_spec=None, params=None, prefix=''):
-            type_checker = wanted
-            wanted = getattr(wanted, '__name__', to_native(type(wanted)))
+        try:
+            type_checker = get_wanted_type(wanted)
```
I think this muddies your API design where the `check_*` methods raise exceptions on error and the `get_*` methods are not exception throwers and always return data instead. So you could either rename it to `check_wanted_type()` or have it return None or empty string for unknown types.

samdoran replied:
That makes sense. I'll make the change.

samdoran committed:
Change get_wanted_type name and behavior

Change the name to get_validator to bette describe what it is doing.

Change the interface to always return a value. This lines up with the behavior of get_*
functions always returning something or None and check_* functions raising an
Exception if something went wrong during the check.

samdoran committed:
Add param to check_type_str()

Not meant to be a long term fix, but gets tests passing. More work is needed to figure
out how to solve this cleanly.

samdoran committed:
Remove private attribute mapping types to validator

Since the function that needs it has moved to parameters.py, there is no need to have it as
a attribute of AnsibleModule.

Update tests that were referencing the private attribute.

samdoran committed:
Use private method for 'str' type

To avoid having to put the string conversion warning behavior in the check_type_str() method,
use the private _check_type_str() method for 'str' type.

Import CHECK_ARGUMENT_TYPES_DISPATCHER for backwards compalitibility and store it as
a private attribute.

Revert changes to support plugins that are referencing serf._CHECK_ARGUMENT_TYPES_DISPATCHER.

samdoran committed:
Add changelog

samdoran committed:
Change function name to better reflect its... function

samdoran committed:
Change dict name to better reflect its contents

CHECK_ARGUMENT_TYPES_DISPATCHER --> DEFAULT_TYPE_VALIDATORS

samdoran committed:
Fix changelog
